### PR TITLE
add system variable tidb_opt_range_max_size

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What is changed, added or deleted? (Required)

Add system variable `tidb_opt_range_max_size`.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v6.3 (TiDB 6.3 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s): https://github.com/pingcap/tidb/issues/37176

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by TomShawn** [comment]
@xuyifangreeneyes Could you please invite a technical reviewer? Thanks!

File: system-variables.md [hunk 1]
```diff
@@ -1807,6 +1807,79 @@ explain select * from t where age=5;
...
 - 指定是否允许优化器将 `Projection` 算子下推到 TiKV 或者 TiFlash。
 
+### `tidb_opt_range_max_size` <span class="version-mark">从 v6.3.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`67108864` (64 MiB)
+- 范围：`[0, 9223372036854775807]`
+- 单位：字节
+- 该变量是优化器构造扫描范围的内存最大限制。当该变量为 `0` 时，表示对扫描范围没有内存限制。当优化器发现构造精确的扫描范围会超出内存限制，优化器会使用更宽松的扫描范围（比如 `[[NULL,+inf]]`）。如果发现执行计划中未使用精确的扫描范围，可以调大该变量让优化器构造精确的扫描范围。
```
Comment (shichun-0415) [inline system-variables.md:1817]:
```suggestion
- 该变量用于指定优化器构造扫描范围的内存最大限制。当该变量为 `0` 时，表示对扫描范围没有内存限制。如果发现构造精确的扫描范围会超出内存限制，优化器会使用更宽松的扫描范围（比如 `[[NULL,+inf]]`）。如果发现执行计划中未使用精确的扫描范围，可以调大该变量让优化器构造精确的扫描范围。
```

File: system-variables.md [hunk 2]
```diff
@@ -1807,6 +1807,117 @@ explain select * from t where age=5;
...
 - 指定是否允许优化器将 `Projection` 算子下推到 TiKV 或者 TiFlash。
 
+### `tidb_opt_range_max_size` <span class="version-mark">从 v6.3.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`67108864` (64 MiB)
+- 范围：`[0, 9223372036854775807]`
+- 单位：字节
+- 该变量用于指定优化器构造扫描范围的内存最大限制。当该变量为 `0` 时，表示对扫描范围没有内存限制。如果发现构造精确的扫描范围会超出内存限制，优化器会使用更宽松的扫描范围（比如 `[[NULL,+inf]]`）。如果发现执行计划中未使用精确的扫描范围，可以调大该变量让优化器构造精确的扫描范围。
```
Comment (TomShawn) [inline system-variables.md:1817]:
```suggestion
- 该变量用于指定优化器构造扫描范围使用的内存上限。当该变量为 `0` 时，表示对扫描范围没有内存限制。如果发现构造精确的扫描范围会超出内存限制，优化器会使用更宽松的扫描范围（例如 `[[NULL,+inf]]`）。如果发现执行计划中未使用精确的扫描范围，可以调大该变量让优化器构造精确的扫描范围。该变量的使用示例如下：
```

File: system-variables.md [hunk 3]
```diff
@@ -1807,6 +1807,117 @@ explain select * from t where age=5;
...
+
```
Comment (TomShawn) [inline system-variables.md:1818]:
```suggestion

查看该变量的默认值：

```

File: system-variables.md [hunk 4]
```diff
@@ -1807,6 +1807,117 @@ explain select * from t where age=5;
...
++----------------------------+
+| @@tidb_opt_range_max_size |
++----------------------------+
+| 67108864                   |
++----------------------------+
+1 row in set (0.01 sec)
+```
+
+```sql
+explain select * from t use index (idx) where a in (10,20,30) and b in (40,50,60);
```
Comment (TomShawn) [inline system-variables.md:1833]:
```suggestion
EXPLAIN SELECT * FROM t use index (idx) WHERE a IN (10,20,30) AND b IN (40,50,60);
```

File: system-variables.md [hunk 5]
```diff
@@ -1807,6 +1807,117 @@ explain select * from t where age=5;
...
+| IndexLookUp_7                 | 0.90    | root      |                          |                                                                                                                                                                             |
+| ├─IndexRangeScan_5(Build)     | 0.90    | cop[tikv] | table:t, index:idx(a, b) | range:[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60], keep order:false, stats:pseudo |
+| └─TableRowIDScan_6(Probe)     | 0.90    | cop[tikv] | table:t                  | keep order:false, stats:pseudo                                                                                                                                              |
++-------------------------------+---------+-----------+--------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
Comment (TomShawn) [inline system-variables.md:1843]: 可以对调整变量值前后，SQL 执行结果给一些对比说明吗？

File: system-variables.md [hunk 6]
```diff
@@ -1807,6 +1807,117 @@ explain select * from t where age=5;
...
 - 指定是否允许优化器将 `Projection` 算子下推到 TiKV 或者 TiFlash。
 
+### `tidb_opt_range_max_size` <span class="version-mark">从 v6.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`67108864` (64 MiB)
+- 取值范围：`[0, 9223372036854775807]`
+- 单位：字节
+- 该变量用于指定优化器构造扫描范围所使用的内存上限。当该变量为 `0` 时，表示对扫描范围没有内存限制。如果发现构造精确的扫描范围会超出内存限制，优化器会使用更宽松的扫描范围（例如 `[[NULL,+inf]]`）。如果发现执行计划中未使用精确的扫描范围，可以调大该变量让优化器构造精确的扫描范围。该变量的使用示例如下：
```
Comment (TomShawn) [inline system-variables.md:1817]:
```suggestion
- 该变量用于指定优化器构造扫描范围的内存用量上限。当该变量为 `0` 时，表示对扫描范围没有内存限制。如果构造精确的扫描范围会超出内存用量限制，优化器会使用更宽松的扫描范围（例如 `[[NULL,+inf]]`）。如果执行计划中未使用精确的扫描范围，可以调大该变量的值让优化器构造精确的扫描范围。该变量的使用示例如下：
```

File: system-variables.md (hunk 5 above)
Comment (xuyifangreeneyes) [reply system-variables.md:1846]: 加了一些说明。

File: system-variables.md [hunk 7]
```diff
@@ -1807,6 +1807,119 @@ explain select * from t where age=5;
...
+| IndexLookUp_7                 | 0.90    | root      |                          |                                                                                                                                                                             |
+| ├─IndexRangeScan_5(Build)     | 0.90    | cop[tikv] | table:t, index:idx(a, b) | range:[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60], keep order:false, stats:pseudo |
+| └─TableRowIDScan_6(Probe)     | 0.90    | cop[tikv] | table:t                  | keep order:false, stats:pseudo                                                                                                                                              |
++-------------------------------+---------+-----------+--------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+3 rows in set (0.00 sec)
+```
+
+```sql
+set @@tidb_opt_range_max_size = 1500;
```
Comment (TomShawn) [inline system-variables.md:1851]:
```suggestion
SET @@tidb_opt_range_max_size = 1500;
```

File: system-variables.md [hunk 8]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+```
+
+```sql
+Query OK, 0 rows affected (0.00 sec)
+```
+
+```sql
+explain select * from t use index (idx) where a in (10,20,30) and b in (40,50,60);
```
Comment (TomShawn) [inline system-variables.md:1859]:
```suggestion
EXPLAIN SELECT * FROM t USE INDEX (idx) WHERE a IN (10,20,30) AND b IN (40,50,60);
```

File: system-variables.md [hunk 9]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
++----------------------------+
+| @@tidb_opt_range_max_size |
++----------------------------+
+| 67108864                   |
++----------------------------+
+1 row in set (0.01 sec)
+```
+
+```sql
+EXPLAIN SELECT * FROM t use index (idx) WHERE a IN (10,20,30) AND b IN (40,50,60);
```
Comment (TomShawn) [inline system-variables.md:1835]: 能否简单解释下这条语句涉及了什么样的扫描范围操作吗？

File: system-variables.md [hunk 10]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+```
+
+在 64MB 的内存最大限制约束下，优化器构造出了精确的 `[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60]` 扫描范围。
```
Comment (TomShawn) [inline system-variables.md:1838]:
```suggestion
在 64 MB 的内存最大限制约束下，优化器构造出了精确的扫描范围 `[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60]`，见如下执行计划返回结果。 
```

File: system-variables.md [hunk 11]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+### `tidb_opt_range_max_size` <span class="version-mark">从 v6.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`67108864` (64 MiB)
+- 取值范围：`[0, 9223372036854775807]`
+- 单位：字节
+- 该变量用于指定优化器构造扫描范围的内存用量上限。当该变量为 `0` 时，表示对扫描范围没有内存限制。如果构造精确的扫描范围会超出内存用量限制，优化器会使用更宽松的扫描范围（例如 `[[NULL,+inf]]`）。如果执行计划中未使用精确的扫描范围，可以调大该变量的值让优化器构造精确的扫描范围。该变量的使用示例如下：
+
+查看该变量的默认值：
```
Comment (TomShawn) [inline system-variables.md:1819]:
```suggestion
查看该变量的默认值，即优化器构造扫描范围最多使用 64 MiB 内存。
```

File: system-variables.md [hunk 12]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+| IndexLookUp_7                 | 0.90    | root      |                          |                                                                                                                                                                             |
+| ├─IndexRangeScan_5(Build)     | 0.90    | cop[tikv] | table:t, index:idx(a, b) | range:[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60], keep order:false, stats:pseudo |
+| └─TableRowIDScan_6(Probe)     | 0.90    | cop[tikv] | table:t                  | keep order:false, stats:pseudo                                                                                                                                              |
++-------------------------------+---------+-----------+--------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+3 rows in set (0.00 sec)
+```
+
```
Comment (TomShawn) [inline system-variables.md:1849]:
```suggestion

现将优化器构造扫描范围的内存用量上限设为 1500 字节。

```

File: system-variables.md [hunk 13]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+```sql
+Query OK, 0 rows affected (0.00 sec)
+```
+
+```sql
+explain select * from t use index (idx) where a in (10,20,30) and b in (40,50,60);
+```
+
+在 1500B 的内存最大限制约束下，优化器构造出了更宽松的 `[10,10], [20,20], [30,30]` 扫描范围，并用 warning 提示用户构造精确的扫描范围需要的内存超出了 `tidb_opt_range_max_size`。
```
Comment (TomShawn) [inline system-variables.md:1862]:
```suggestion
在 1500 字节内存的最大限制约束下，优化器构造出了更宽松的扫描范围 `[10,10], [20,20], [30,30]`，并用 warning 提示用户构造精确的扫描范围所需的内存用量超出了 `tidb_opt_range_max_size` 的限制。

```

File: system-variables.md [hunk 14]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+| ├─Selection_7(Build)          | 0.09    | cop[tikv] |                          | in(test.t.b, 40, 50, 60)                                        |
+| │ └─IndexRangeScan_5          | 30.00   | cop[tikv] | table:t, index:idx(a, b) | range:[10,10], [20,20], [30,30], keep order:false, stats:pseudo |
+| └─TableRowIDScan_6(Probe)     | 0.09    | cop[tikv] | table:t                  | keep order:false, stats:pseudo                                  |
++-------------------------------+---------+-----------+--------------------------+-----------------------------------------------------------------+
+4 rows in set, 1 warning (0.00 sec)
+```
+
+
+```sql
+show warnings;
```
Comment (TomShawn) [inline system-variables.md:1877]:
```suggestion
SHOW WARNINGS;
```

File: system-variables.md [hunk 15]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+
+```sql
++---------+------+---------------------------------------------------------------------------------------------------------------------------------------------+
+| Level   | Code | Message                                                                                                                                     |
++---------+------+---------------------------------------------------------------------------------------------------------------------------------------------+
+| Warning | 1105 | Memory capacity of 1500 bytes for 'tidb_opt_range_max_size' exceeded when building ranges. Less accurate ranges such as full range are chosen |
++---------+------+---------------------------------------------------------------------------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
```
Comment (TomShawn) [inline system-variables.md:1888]:
```suggestion

再将优化器构造扫描范围的内存用量上限设为 100 字节。

```

File: system-variables.md [hunk 16]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+```sql
+set @@tidb_opt_range_max_size = 100;
+```
+
+```sql
+Query OK, 0 rows affected (0.00 sec)
+```
+
+```sql
+explain select * from t use index (idx) where a in (10,20,30) and b in (40,50,60);
```
Comment (TomShawn) [inline system-variables.md:1898]:
```suggestion
EXPLAIN SELECT * FROM t USE INDEX (idx) WHERE a IN (10,20,30) AND b IN (40,50,60);
```

File: system-variables.md [hunk 17]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+```
+
+在 100B 的内存最大限制约束下，优化器选择了 `IndexFullScan`，并用 warning 提示用户构造精确的扫描范围需要的内存超出了 `tidb_opt_range_max_size`。
```
Comment (TomShawn) [inline system-variables.md:1901]:
```suggestion
在 100 字节的内存最大限制约束下，优化器选择了 `IndexFullScan`，并用 warning 提示用户构造精确的扫描范围所需的内存超出了 `tidb_opt_range_max_size` 的限制。

```

File: system-variables.md [hunk 18]
```diff
@@ -1807,6 +1807,123 @@ explain select * from t where age=5;
...
+| IndexLookUp_8                 | 8000.00  | root      |                          |                                                    |
+| ├─Selection_7(Build)          | 8000.00  | cop[tikv] |                          | in(test.t.a, 10, 20, 30), in(test.t.b, 40, 50, 60) |
+| │ └─IndexFullScan_5           | 10000.00 | cop[tikv] | table:t, index:idx(a, b) | keep order:false, stats:pseudo                     |
+| └─TableRowIDScan_6(Probe)     | 8000.00  | cop[tikv] | table:t                  | keep order:false, stats:pseudo                     |
++-------------------------------+----------+-----------+--------------------------+----------------------------------------------------+
+4 rows in set, 1 warning (0.00 sec)
+```
+
+```sql
+show warnings;
```
Comment (TomShawn) [inline system-variables.md:1915]:
```suggestion
SHOW WARNINGS;
```

File: system-variables.md (hunk 9 above)
Comment (xuyifangreeneyes) [reply system-variables.md:1835]: 在索引 `idx` 上的精确扫描范围是 `[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60]`。

File: system-variables.md [hunk 19]
```diff
@@ -1807,6 +1807,129 @@ explain select * from t where age=5;
...
+| 67108864                   |
++----------------------------+
+1 row in set (0.01 sec)
+```
+
+```sql
+EXPLAIN SELECT * FROM t use index (idx) WHERE a IN (10,20,30) AND b IN (40,50,60);
+```
+
+在 64 MB 的内存最大限制约束下，优化器构造出了精确的扫描范围 `[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60]`，见如下执行计划返回结果。
```
Comment (TomShawn) [inline system-variables.md:1838]:
```suggestion
在 64 MiB 的内存最大限制约束下，优化器构造出精确的扫描范围 `[10 40,10 40], [10 50,10 50], [10 60,10 60], [20 40,20 40], [20 50,20 50], [20 60,20 60], [30 40,30 40], [30 50,30 50], [30 60,30 60]`，见如下执行计划返回结果。 

```

**Review by TomShawn — APPROVED** [review]
LGTM

File: system-variables.md [hunk 20]
```diff
@@ -1807,6 +1807,137 @@ explain select * from t where age=5;
...
+| id                            | estRows | task      | access object            | operator info                                                   |
++-------------------------------+---------+-----------+--------------------------+-----------------------------------------------------------------+
+| IndexLookUp_8                 | 0.09    | root      |                          |                                                                 |
+| ├─Selection_7(Build)          | 0.09    | cop[tikv] |                          | in(test.t.b, 40, 50, 60)                                        |
+| │ └─IndexRangeScan_5          | 30.00   | cop[tikv] | table:t, index:idx(a, b) | range:[10,10], [20,20], [30,30], keep order:false, stats:pseudo |
+| └─TableRowIDScan_6(Probe)     | 0.09    | cop[tikv] | table:t                  | keep order:false, stats:pseudo                                  |
++-------------------------------+---------+-----------+--------------------------+-----------------------------------------------------------------+
+4 rows in set, 1 warning (0.00 sec)
+```
+
```
Comment (TomShawn) [inline system-variables.md:1883]:
```suggestion
```

**Review by hfxsd — APPROVED** [review]
LGTM

**Review by chrysan — APPROVED** [review]

**Pull request merged**